mm2src/coins/coin_balance.rs

```diff
@@ -0,0 +1,452 @@
+pub struct HDAccountBalance {
+    pub account_index: u32,
+    pub derivation_path: RpcDerivationPath,
+    pub total_balance: CoinBalance,
+    pub addresses: Vec<HDAddressBalance>,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize)]
+pub struct HDAddressBalance {
+    pub address: String,
+    pub derivation_path: RpcDerivationPath,
+    pub chain: Bip44Chain,
+    pub balance: CoinBalance,
+}
+
+#[derive(Deserialize)]
+pub struct HDAccountBalanceRequest {
+    coin: String,
+    #[serde(flatten)]
+    params: HDAccountBalanceParams,
+}
+
+#[derive(Deserialize)]
+pub struct HDAccountBalanceParams {
+    pub account_index: u32,
+    pub chain: Bip44Chain,
+    #[serde(default = "common::ten")]
+    pub limit: usize,
+    #[serde(default)]
+    pub paging_options: PagingOptionsEnum<u32>,
+}
+
+#[derive(Deserialize)]
+pub struct CheckHDAccountBalanceRequest {
+    coin: String,
+    #[serde(flatten)]
+    params: CheckHDAccountBalanceParams,
+}
+
+#[derive(Deserialize)]
+pub struct CheckHDAccountBalanceParams {
+    pub account_index: u32,
+    pub gap_limit: Option<u32>,
+}
+
+#[derive(Debug, PartialEq, Serialize)]
+pub struct HDAccountBalanceResponse {
+    pub account_index: u32,
+    pub derivation_path: RpcDerivationPath,
+    pub addresses: Vec<HDAddressBalance>,
+    pub limit: usize,
+    pub skipped: u32,
+    pub total: u32,
+    pub total_pages: usize,
+    pub paging_options: PagingOptionsEnum<u32>,
+}
+
+#[derive(Debug, PartialEq, Serialize)]
+pub struct CheckHDAccountBalanceResponse {
+    pub account_index: u32,
+    pub derivation_path: RpcDerivationPath,
+    pub new_addresses: Vec<HDAddressBalance>,
+}
+
+#[async_trait]
+pub trait EnableCoinBalanceOps {
+    async fn enable_coin_balance(&self) -> BalanceResult<EnableCoinBalance>;
+}
+
+#[async_trait]
+impl<Coin, Address, HDWallet, HDAccount, HDAddressChecker> EnableCoinBalanceOps for Coin
+where
+    Coin: AddressBalanceOps<Address = Address>
+        + CoinWithDerivationMethod<Address = Address, HDWallet = HDWallet>
+        + HDWalletBalanceOps<HDWallet = HDWallet, HDAccount = HDAccount, HDAddressChecker = HDAddressChecker>
+        + Sync,
+    Address: fmt::Display + Sync,
+    HDWallet: Sync,
+    HDAddressChecker: HDAddressBalanceChecker,
+{
+    async fn enable_coin_balance(&self) -> BalanceResult<EnableCoinBalance> {
+        match self.derivation_method() {
+            DerivationMethod::Iguana(address) => self.address_balance(address).await.map(|balance| {
+                EnableCoinBalance::Iguana(IguanaWalletBalance {
+                    address: address.to_string(),
+                    balance,
+                })
+            }),
+            DerivationMethod::HDWallet(hd_wallet) => self
+                .enable_hd_wallet_balance(hd_wallet)
+                .await
+                .map(EnableCoinBalance::HD),
+        }
+    }
+}
+
+#[async_trait]
+pub trait HDWalletBalanceOps: AddressBalanceOps {
+    type HDWallet;
```

Comment:
Shouldn't `HDWallet` be bound to a trait, which has `HDAccount` and `HDAddressChecker` associated types?

Reply:
Done
Now it looks better. Thanks for the tip!